### PR TITLE
Deliver full intersection type of a value to the IDE

### LINKED ISSUE #11481: Deliver full intersection type of a value to the IDE
To provide 
- #11366

the IDE needs to receive information about _all types_ participating in an **intersection type** of a value. Right now the IDE receives only one of the types in the **intersection type**. That has to be changed, probably by adjusting API or implementation of [ExpressionUpdate](https://github.com/enso-org/enso/blob/2bbd90970547eed11b4c52f5b73c0bac785379f3/docs/language-server/protocol-language-server.md#expressionupdate) message.

## Specification

Imagine [following program](https://github.com/orgs/enso-org/discussions/11366#discussioncomment-11139648):
```ruby
from Standard.Base import all
import Standard.Visualization

type R
    private V a:Text i:Integer

    as_text self = self:Text
    as_int_text self = 
        r = (self : Integer & Text)
        r
    as_text_int self = 
        r = (self : Text & Integer)
        r
    as_int self = self:Integer

Text.from (that:R) = that.a
Integer.from (that:R) = that.i

main =
    n = R.V "Hi" 3
    t1 = n.as_text
    b2 = n.as_int_text
    i3 = n.as_int
    node1 = n.as_text_int
```
Right now (after #11480 is integrated) the type of `b2` is `Integer` and type of `node1` is `Text`.

## Goal

The type of the `b2` in the above example should include `Integer` as well as `Text` and the type of `node1` should include `Text` as well as `Integer`.

#### Why?

Only then the IDE can offer methods for both types in the component browser for `b2` or `node1`

#### Note

To support **Auto-Domination** use-case of #11366  - the order of types may be important.

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

close #11481 

Changelog:
- update: handle `MultiTypeValue` results in the execution instrument
- update: language server protocol supports multi-type values
- update: GUI uses only the first type of a multi-type value when 

### Important Notes

GUI uses only the first type of the intersection. See the difference between `Integer&Text` and `Text&Integer`:

https://github.com/user-attachments/assets/29efc89b-c223-4043-8dff-9cdae1987f0c

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

File: app/ydoc-shared/src/languageServerTypes.ts [hunk 1]
```diff
@@ -195,7 +195,7 @@ export interface ExpressionUpdate {
   /** The id of updated expression. */
   expressionId: ExpressionId
   /** The updated type of the expression. */
-  type?: string
+  type: string[]
```
Comment (JaroslavTulach) [inline app/ydoc-shared/src/languageServerTypes.ts:198]: This is an incompatible message change. Single values used to be represented as `"string"` and now they will be `[ "string" ]`. Are we sure we don't need compatibility?

File: docs/language-server/protocol-language-server.md [hunk 2]
```diff
@@ -367,7 +367,7 @@ interface ExpressionUpdate {
   /** The id of updated expression. */
   expressionId: ExpressionId;
   /** The updated type of the expression. */
-  type?: string;
+  type: string[];
```
Comment (JaroslavTulach) [inline docs/language-server/protocol-language-server.md:376]: I feel incompatible changes should be versioned and documented. We should:
- increase the version of the protocol
- describe what change has happened and why it is incompatible


Sure, I know we have no version of the protocol. However I asked for it a year ago. Moreover recently there were complains about incompatibilities between cloud/IDE and the conclusion was: we should at least detect the incompatibilities by versioning our protocols.

File: engine/polyglot-api/src/main/scala/org/enso/polyglot/runtime/Runtime.scala [hunk 3]
```diff
@@ -108,20 +108,18 @@ object Runtime {
...
+      * @param fromCache whether the value for this expression came from the cache
+      * @param typeChanged whether the type of the value or method definition
       * has changed from the one that was cached, if any
       * @param payload an extra information about the computed value
       */
     @named("expressionUpdate")
     case class ExpressionUpdate(
       expressionId: ExpressionId,
-      expressionType: Option[String],
+      expressionTypes: Option[Vector[String]],
```
Comment (JaroslavTulach) [inline engine/polyglot-api/src/main/scala/org/enso/polyglot/runtime/Runtime.scala:122]: That's not the right type as in the JSON:
- the vector may be empty
- but it must be present

File: engine/runtime-instrument-common/src/main/java/org/enso/interpreter/service/ExecutionCallbacks.java [hunk 4]
```diff
@@ -214,20 +215,29 @@ private FunctionCallInfo functionCallInfoById(UUID nodeId) {
...
-      var typeOfNode = TypeOfNode.getUncached();
-      Object typeResult = value == null ? null : typeOfNode.execute(value);
-      if (typeResult instanceof Type t) {
-        resultType = getTypeQualifiedName(t);
-      } else {
-        resultType = null;
+      return new String[] {Constants.UNRESOLVED_SYMBOL};
+    }
+
+    if (value instanceof EnsoMultiValue multiValue) {
```
Comment (JaroslavTulach) [inline engine/runtime-instrument-common/src/main/java/org/enso/interpreter/service/ExecutionCallbacks.java]: Ideally this should be encapsulated in `TypeOfNode`. Having these extra checks for `EnsoMultiValue` and `UnresolvedSymbol` here will just cause inconsistency in the future.

File: engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/ProgramExecutionSupport.scala [hunk 5]
```diff
@@ -198,14 +198,14 @@ object ProgramExecutionSupport {
         val notExecuted =
           methodCallsCache.getNotExecuted(executionFrame.cache.getCalls)
         notExecuted.forEach { expressionId =>
-          val expressionType = executionFrame.cache.getType(expressionId)
-          val expressionCall = executionFrame.cache.getCall(expressionId)
+          val expressionTypes = executionFrame.cache.getType(expressionId)
```
Comment (JaroslavTulach) [inline engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/ProgramExecutionSupport.scala:201]: Shouldn't the `getType` method be renamed to `getTypes`?

File: engine/runtime-integration-tests/src/test/scala/org/enso/interpreter/test/instrument/RuntimeIntersectionTypesTest.scala [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+import org.scalatest.BeforeAndAfterEach
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+
+import java.io.{ByteArrayOutputStream, File}
+import java.nio.file.{Files, Paths}
+import java.util.UUID
+
+@scala.annotation.nowarn("msg=multiarg infix syntax")
+class RuntimeIntersectionTypesTest
```
Comment (JaroslavTulach) [inline engine/runtime-integration-tests/src/test/scala/org/enso/interpreter/test/instrument/RuntimeIntersectionTypesTest.scala]: Ech. New Scala file.

**Review by JaroslavTulach — APPROVED** [review]
The change itself is fine, but it reveals long term unresolved issues:
- are incompatible changes in the protocol OK?
- versioning on the protocol is missing
- duplication of functionality instead of encapsulating it in `TypeOfNode`

File: docs/language-server/protocol-language-server.md [hunk 7]
```diff
@@ -367,7 +367,7 @@ interface ExpressionUpdate {
   /** The id of updated expression. */
   expressionId: ExpressionId;
   /** The updated type of the expression. */
```
Comment (kazcw) [inline docs/language-server/protocol-language-server.md]: The current docs are not clear about some details of the protocol. An empty array seems to be treated specially; judging from the implementation, it looks like:
- When the status is `Pending`, an empty array can indicate that the type has not been determined yet.
- When the status is not `Pending`, an empty array indicates that no specific type is known for the value--maybe this can happen either if the type is known to be `Any`. Can this also happen if the type could not be determined?

Are these updates incremental? It's not clear whether in the pending state `[]` might be sent to indicate "no change", or every update will contain all currently-known type information even if it hasn't changed.

**Review by kazcw — APPROVED** [review]

File: engine/language-server/src/main/scala/org/enso/languageserver/search/SuggestionsHandler.scala [hunk 8]
```diff
@@ -264,54 +262,6 @@ final class SuggestionsHandler(
         )
       )
 
-    case msg: Api.ExpressionUpdates if state.isSuggestionUpdatesRunning =>
-      state.suggestionUpdatesQueue.enqueue(msg)
-
-    case Api.ExpressionUpdates(_, updates) =>
```
Comment (hubertp) [inline engine/language-server/src/main/scala/org/enso/languageserver/search/SuggestionsHandler.scala:270]: Why this change? Seems rather unrelated. Or is that feature no longer used?

File: engine/polyglot-api/src/main/scala/org/enso/polyglot/runtime/Runtime.scala (hunk 3 above)
Comment (hubertp) [reply engine/polyglot-api/src/main/scala/org/enso/polyglot/runtime/Runtime.scala:122]: Isn't that the point of this change? 
```
expression-types: null
```
(when `None`)
and
```
expression-types: []
```
are both allowed in JSON and yet represent two different scenarios and should be handled separately?
But maybe I'm missing something, @4e6 ?

**Review by hubertp — COMMENTED** [review]
In this particular case I would agree with Jaroslav that the binary incompatible change this PR introduces might be problematic.
We could have both, `type` and `types`, fields be present for the time being at the cost of some duplication. Then GUI could make a switch to support intersection types more gently and we could deprecate (and eventually remove)`type`.

File: docs/language-server/protocol-language-server.md (hunk 7 above)
Comment (4e6) [reply docs/language-server/protocol-language-server.md]: I updated the docs. The logic did not change. The array is just a way to encode intersection types
- `[]` indicates no type information (previously `null`)
- `[T]` represent a type of the expression (previously `T`)
- `[T,U]` represents an intersection type `T & U` (previously not supported, probably delivered just as a `T` type)

File: engine/language-server/src/main/scala/org/enso/languageserver/search/SuggestionsHandler.scala (hunk 8 above)
Comment (4e6) [reply engine/language-server/src/main/scala/org/enso/languageserver/search/SuggestionsHandler.scala:270]: Yes, after the component browser is finished, and GUI can handle suggestions itself, we don't need to do this logic anymore

File: engine/polyglot-api/src/main/scala/org/enso/polyglot/runtime/Runtime.scala (hunk 3 above)
Comment (4e6) [reply engine/polyglot-api/src/main/scala/org/enso/polyglot/runtime/Runtime.scala:122]: Jaroslav's point was that an empty vector is the same as `None`. But because it is a valid state in the engine, I made it this way to transfer the raw engine response to the language server. In the user-facing update message the type is `type: string[]`

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/library/dispatch/TypeOfNode.java [hunk 9]
```diff
@@ -75,6 +75,11 @@ Object doWarning(WithWarnings value, @Cached TypeOfNode withoutWarning) {
     return withoutWarning.execute(value.getValue());
   }
 
+  @Specialization
+  Object doEnsoMultiValue(EnsoMultiValue value) {
+    return value.allTypes();
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/library/dispatch/TypeOfNode.java]: Hello Dmitry,
thanks for trying to implement my request to concentrate the logic of obtaining a type array into `TypeOfNode` in 86bfb986d185b4a01104753a20c9987591d855f3, but this cannot work - the `Object` here means: [any valid interop value](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/InteropLibrary.html#isValidValue(java.lang.Object)) and Java arrays aren't such values.

Such a code change isn't likely to pass the **Truffle** `assert` checks as for example this [failure indicates](https://github.com/enso-org/enso/actions/runs/11942161547/job/33288566047?pr=11583#step:7:3194)

The change has to be done differently. I'll prepare a PR for review:
- #11618

With [87e738b](https://github.com/enso-org/enso/pull/11618/commits/87e738b77981bf419174ffc4b2caea74bc8315a7) there is `findAllTypes` method that provides all the `Type[]` associated with a value.

File: engine/polyglot-api/src/main/scala/org/enso/polyglot/runtime/Runtime.scala (hunk 3 above)
Comment (JaroslavTulach) [reply engine/polyglot-api/src/main/scala/org/enso/polyglot/runtime/Runtime.scala:122]: I see. I still have problems to properly identify the _"JSON protocol level"_ each class represents. I am overwhelmed by _too many JSON messages_ situation.

**Review by JaroslavTulach — CHANGES_REQUESTED** [review]
The change in `TypeOfNode` isn't correct and fails on **Truffle asserts**. I'll prepare an alternative approach.

**Comment by somebody1234** [comment]
[mockSuggestions.json](https://github.com/enso-org/enso/blob/99cca042f210367376b643cded5c18730ded4304/app/gui/src/project-view/mock/mockSuggestions.json) probably needs to be regenerated here (@farmaazon?)

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/library/dispatch/TypeOfNode.java (hunk 9 above)
Comment (4e6) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/library/dispatch/TypeOfNode.java]: I can wrap the result in `EnsoObject`

Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/library/dispatch/TypeOfNode.java]: > I can wrap the result in EnsoObject

Interesting idea, yes that would fix the **Truffle assertion errors**:
- use `ArrayLikeHelpers.wrapEnsoObjects` 
- let's see what other failures we get

**Comment by 4e6** [comment]
@somebody1234 it does not change the suggestion interface. This change is only about the expression update.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/library/dispatch/TypeOfNode.java (hunk 9 above)
Comment (4e6) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/library/dispatch/TypeOfNode.java]: I guess, I should wait for #11618

Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/library/dispatch/TypeOfNode.java]: - yes, use #11618 API now when it has been integrated

File: engine/runtime-instrument-common/src/main/java/org/enso/interpreter/service/ExecutionCallbacks.java [hunk 10]
```diff
@@ -216,12 +216,24 @@ private FunctionCallInfo functionCallInfoById(UUID nodeId) {
...
-      Object typeResult = value == null ? null : typeOfNode.findTypeOrError(value);
-      if (typeResult instanceof Type t) {
-        resultType = getTypeQualifiedName(t);
-      } else {
-        resultType = null;
+      return new String[] {Constants.UNRESOLVED_SYMBOL};
+    }
+
+    var typeOfNode = TypeOfNode.getUncached();
+    Type[] allTypes = value == null ? null : typeOfNode.findAllTypesOrNull(value);
```
Comment (JaroslavTulach) [inline engine/runtime-instrument-common/src/main/java/org/enso/interpreter/service/ExecutionCallbacks.java:223]: Nice.

**Review by JaroslavTulach — APPROVED** [review]
